packages/grafana-data/src/utils/vector.ts

```diff
@@ -109,18 +99,101 @@ export class CircularVector<T = any> implements Vector<T> {
   }
 }
 
+interface CircularOptions<T> {
+  buffer?: T[];
+  append?: 'head' | 'tail';
+  capacity?: number;
+}
+
 /**
- * Values are returned in the order defined by the input parameter
+ * Circular vector uses a single buffer to capture a stream of values
+ * overwriting the oldest value on add.
+ *
+ * This supports addting to the 'head' or 'tail' and will grow the buffer
+ * to match a configured capacity.
  */
-export class SortedVector<T = any> implements Vector<T> {
-  constructor(private source: Vector<T>, private order: number[]) {}
+export class CircularVector<T = any> implements Vector<T> {
+  private buffer: T[];
+  private index: number;
+  private capacity: number;
+  private tail: boolean;
+
+  constructor(options: CircularOptions<T>) {
+    this.buffer = options.buffer || [];
+    this.capacity = this.buffer.length;
+    this.tail = 'head' !== options.append;
+    this.index = 0;
 
-  get length(): number {
-    return this.source.length;
+    this.add = this.getAddFunction();
+    if (options.capacity) {
+      this.setCapacity(options.capacity);
+    }
   }
 
-  get(index: number): T {
-    return this.source.get(this.order[index]);
+  // Pick an add function that either writes forwards or backwards
+  private getAddFunction() {
+    return this.tail
+      ? (value: T) => {
+          this.buffer[this.index] = value;
+          this.index = (this.index + 1) % this.buffer.length;
+        }
+      : (value: T) => {
+          let idx = this.index - 1;
+          if (idx < 0) {
+            idx = this.buffer.length - 1;
+          }
+          this.buffer[idx] = value;
+          this.index = idx;
+        };
+  }
+
+  setCapacity(v: number) {
+    if (this.capacity === v) {
+      return;
+    }
+    // Make a copy so it is in order and new additions can be at the head or tail
+    const copy = this.toArray();
+    if (v > this.length) {
+      this.buffer = copy;
+
+      // Change the 'add' function so it actually appends
+      this.add = this.tail
+        ? (value: T) => {
```

Comment:
nit pick / personal opinion, I am not in love with using ? for multi-line expressions as I think it hurts readability.  Curious is the reason for making the add function dynamic like this performance? (to not have to check direction on every add). There is a duplicate this.tail ? expression in the getAddFunction which looks a bit strange (in having sort of two nested conditions that does check the same thing).